src/openrct2/core/OrcaStream.hpp

```diff
@@ -0,0 +1,679 @@
+                    if constexpr (std::is_signed<T>())
+                    {
+                        int32_t raw{};
+                        Read(&raw, sizeof(raw));
+                        if (raw < std::numeric_limits<T>::min() || raw > std::numeric_limits<T>::max())
+                        {
+                            throw std::runtime_error("Value is incompatible with internal type.");
+                        }
+                        return static_cast<T>(raw);
+                    }
+                    else
+                    {
+                        uint32_t raw{};
+                        Read(&raw, sizeof(raw));
+                        if (raw > std::numeric_limits<T>::max())
+                        {
+                            throw std::runtime_error("Value is incompatible with internal type.");
+                        }
+                        return static_cast<T>(raw);
+                    }
+                }
+            }
+
+            template<typename T, typename = std::enable_if<std::is_integral<T>::value>> void WriteInteger(const T value)
+            {
+                if constexpr (sizeof(T) > 4)
+                {
+                    if constexpr (std::is_signed<T>())
+                    {
+                        auto raw = static_cast<int64_t>(value);
+                        Write(&raw, sizeof(raw));
+                    }
+                    else
+                    {
+                        auto raw = static_cast<uint64_t>(value);
+                        Write(&raw, sizeof(raw));
+                    }
+                }
+                else
+                {
+                    if constexpr (std::is_signed<T>())
+                    {
+                        auto raw = static_cast<int32_t>(value);
+                        Write(&raw, sizeof(raw));
+                    }
+                    else
+                    {
+                        auto raw = static_cast<uint32_t>(value);
+                        Write(&raw, sizeof(raw));
+                    }
+                }
+            }
+
+            std::string ReadString()
+            {
+                std::string buffer;
+                buffer.reserve(64);
+                while (true)
+                {
+                    char c;
+                    ReadBuffer(&c, sizeof(c));
+                    if (c == 0)
+                    {
+                        break;
+                    }
+                    buffer.push_back(c);
+                }
+                buffer.shrink_to_fit();
+                return buffer;
+            }
+
+            void WriteString(const std::string_view s)
+            {
+                const char nullt = '\0';
+                auto len = s.find('\0');
+                if (len == std::string_view::npos)
+                {
+                    len = s.size();
+                }
+                _buffer.Write(s.data(), len);
+                _buffer.Write(&nullt, sizeof(nullt));
```

Comment:
```suggestion
                const char nullt = '\0';
                auto len = s.size();
                while (len > 0 && s[len - 1] == nullt)
                {
                    len--;
                }
```
I think in 99.9% of the time this will simply skip 1 byte, not sure if this is cleaner to read but we don't have to search for null terminators at the beginning.